### DISCUSSION TITLE
how does mysql.createPool work?

### DISCUSSION THREAD

**Discussion by MassMessage** [sidorares/node-mysql2#1747] [General]
how does  `mysql.createPool` work? let's say i have the following code:

```typscript
const getConn = () => {
    const connection = mysql.createPool({
        host: process.env.DB_HOST,
        user: process.env.DB_USER,
        password: process.env.DB_PASS,
        database: process.env.DB_NAME
    });
    return connection;
}

const foo = () => {
    const coon = getConn();
    coon.query('select * from YYY',
        function(err, results, fields) {
            console.log(results);
        }
    );
}


const baa = () => {
    const coon = getConn();
    coon.query('select * from ZZZ',
        function(err, results, fields) {
            // do something
        }
    );
}
```

i do call `foo` and later on `baa`. in the call `foo` the connection is created and in `baa` and subsequents call to `getConn` i'll get just the same connection instance created by `foo`, is that right?

**Comment by sidorares** [comment]
you are creating a new pool instance in `getConn`, which manages its own set of connections. In your example connections are not shared or reused.

What you should do instead ( I'm using async api but callback version is similar )

```js

const pool = mysql.createPool(connectOptions);

const foo = async () => {
  const conn = await pool.getConnection();
  const [rows] = conn.query('select * from YYY');
  conn.release();
  console.log(rows);
};

const baa = async () => {
  // or just do .query() on a pool, it'll do .getConnection + query + release internally
  const [rows] = pool.query('select * from YYY');
  console.log(rows);
};

```

**Reply by MassMessage** [reply]
> you are creating a new pool instance in `getConn`, which manages its own set of connections. In your example connections are not shared or reused.
> 
> What you should do instead ( I'm using async api but callback version is similar )
> 
> ```js
> const pool = mysql.createPool(connectOptions);
> 
> const foo = async () => {
>   const conn = await pool.getConnection();
>   const [rows] = conn.query('select * from YYY');
>   conn.release();
>   console.log(rows);
> };
> 
> const baa = async () => {
>   // or just do .query() on a pool, it'll do .getConnection + query + release internally
>   const [rows] = pool.query('select * from YYY');
>   console.log(rows);
> };
> ```

I"ve tried to make a version without callbacks but i'm getting error from query(). code:

```
const getName2 =  (conn: mysql.PoolConnection, code:number): string | null =>
{
    const [results] = conn.query('select bares_nome from bares where bares_cod = ?', [code]);
        
    if(IsRowDataPacket(results) && results.length === 0)
    { 
        return null;
    }

    interface IBar { bares_nome: string; }
    const data = results as IBar[];
    const r:IBar = data[0];
    return r.bares_nome;
}
```

but on `results` var i'm getting the error:

`Type 'Query' is not an array type.ts(2461)`

**Reply by sidorares** [reply]
is that just a types error ( e.i does it work as expected without a type check )?

**Reply by MassMessage** [reply]
> is that just a types error ( e.i does it work as expected without a type check )?

it doesn't compile.  I get the error:

> Type error: Type 'Query' is not an array type.
